ffmpeg_new.c

```diff
@@ -0,0 +1,555 @@
+	uint8_t *buffer;
+	uint8_t *buffer_malloc;
+	uint8_t *buffer_pos;	/* current buffer position */
+	int buffer_used_len;
+};
+
+struct ffmpeg_private {
+	AVCodecContext *codec_context;
+	AVFormatContext *input_context;
+	AVCodec *codec;
+	SwrContext *swr;
+
+	struct ffmpeg_input *input;
+	struct ffmpeg_output *output;
+};
+
+#if LIBAVCODEC_VERSION_INT < AV_VERSION_INT(57, 40, 100)
+static int ffmpeg_receive_frame(AVCodecContext *cc, AVFrame *frame)
+{
+	struct ffmpeg_private *private = cc->opaque;
+	struct ffmpeg_input *input = private->input;
+
+	int len;
+	int got_frame;
+
+ 	if (input->curr_pkt_size <= 0) {
+		return -1;
+ 	}
+ 
+ 	AVPacket avpkt;
+ 	av_new_packet(&avpkt, input->curr_pkt_size);
+ 	memcpy(avpkt.data, input->curr_pkt_buf, input->curr_pkt_size);
+ 	len = avcodec_decode_audio4(cc, frame, &got_frame, &avpkt);
+ 	ffmpeg_packet_unref(&avpkt);
+
+	if (len < 0) {
+		input->curr_pkt_size = 0;
+		return -1;
+	}
+
+	input->curr_pkt_size -= len;
+	input->curr_pkt_buf += len;
+
+	return !got_frame;
+}
+
+static int ffmpeg_send_packet(AVCodecContext *cc, AVPacket *pkt)
+{
+	struct ffmpeg_private *private = cc->opaque;
+	struct ffmpeg_input *input = private->input;
+
+	if (pkt) {
+		input->curr_pkt_size = pkt->size;
+		input->curr_pkt_buf = pkt->data;
+	}
+
+	return 0;
+}
+#endif
+
+static struct ffmpeg_input *ffmpeg_input_create(void)
+{
+	struct ffmpeg_input *input = xnew(struct ffmpeg_input, 1);
+
+	if (av_new_packet(&input->pkt, 0) != 0) {
+		free(input);
+		return NULL;
+	}
+	input->curr_pkt_size = 0;
+	input->curr_pkt_buf = input->pkt.data;
+	input->eof = false;
+	return input;
+}
+
+static void ffmpeg_input_free(struct ffmpeg_input *input)
+{
+	ffmpeg_packet_unref(&input->pkt);
+	free(input);
+}
+
+static struct ffmpeg_output *ffmpeg_output_create(void)
+{
+	struct ffmpeg_output *output = xnew(struct ffmpeg_output, 1);
+
+	output->buffer_malloc = xnew(uint8_t, AVCODEC_MAX_AUDIO_FRAME_SIZE + 15);
+	output->buffer = output->buffer_malloc;
+	/* align to 16 bytes so avcodec can SSE/Altivec/etc */
+	while ((intptr_t) output->buffer % 16)
+		output->buffer += 1;
+	output->buffer_pos = output->buffer;
+	output->buffer_used_len = 0;
+	return output;
+}
+
+static void ffmpeg_output_free(struct ffmpeg_output *output)
+{
+	free(output->buffer_malloc);
+	output->buffer_malloc = NULL;
+	output->buffer = NULL;
+	free(output);
+}
+
+static inline void ffmpeg_buffer_flush(struct ffmpeg_output *output)
+{
+	output->buffer_pos = output->buffer;
+	output->buffer_used_len = 0;
+}
+
+static void ffmpeg_init(void)
+{
+	static int inited = 0;
+
+	if (inited != 0)
+		return;
+	inited = 1;
+
+	av_log_set_level(AV_LOG_QUIET);
+
+	/* We could register decoders explicitly to save memory, but we have to
+	 * be careful about compatibility. */
+	av_register_all();
+}
+
+static int ffmpeg_open(struct input_plugin_data *ip_data)
+{
+	struct ffmpeg_private *priv;
+	int err = 0;
+	int stream_index = -1;
+	int64_t channel_layout = 0;
+	AVCodec *codec;
```

Comment:
`= NULL` as well, for consistency.

I'd say it is almost always better to explicitly initialize variables. This way, reader of the code does not need to double-check «is this always properly initialized later?» all the time.